### ISSUE TITLE
Token Cache Does not appear to be User Specific

### ISSUE THREAD

**Issue by jondmcelroy** [AzureAD/microsoft-authentication-library-for-dotnet#725]
**Which Version of MSAL are you using ?**
2.3.1-preview

**Which platform has the issue?**
dotnetcore

**What authentication flow has the issue?**
* Desktop / Mobile
    * [ ] Interactive
    * [ ] Integrated Windows Auth
    * [ ] Username Password
    * [ ] Device code flow (browserless)
* Web App
    * [X] Authorization code
    * [ ] OBO
* Web API
    * [ ] OBO

Other? - please describe;

**Repro**
* I am using B2C for authentication. 
* I followed the examples for setting up the OpenIdConnections here: https://github.com/Azure-Samples/active-directory-b2c-dotnetcore-webapp/blob/master/WebApp-OpenIDConnect-DotNet/OpenIdConnectOptionsSetup.cs
* I setup my application with the MSAL Session Cache token examples from here: https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/wiki/token-cache-serialization#token-cache-for-a-web-app-confidential-client-application
* I am using the examples for how to use AcquireTokenSilentAsync from here: https://azure.microsoft.com/en-us/resources/samples/active-directory-b2c-dotnetcore-webapp/
* Final project looks similar to the examples here: https://github.com/Azure-Samples/active-directory-b2c-dotnet-webapp-and-webapi

The Application is secured using B2C authentication and then retrieves token to use for making API calls to a different server.
Here is my code to created a ConfidentialClientApplication and call AcquireTokenSilentAsync to get a token for the currently logged in user.

```csharp
var signedInUserId = HttpContext.User.FindFirst(AzureAdB2COptions.ObjectIdentifierClaim)?.Value;
var userTokenCache = new MsalSessionCache(signedInUserId, HttpContext).GetMsalCacheInstance();
var app = new ConfidentialClientApplication(
    _azureAdB2COptions.ClientId,
    _azureAdB2COptions.Authority,
    _azureAdB2COptions.RedirectUri,
    new ClientCredential(_azureAdB2COptions.ClientSecret),
    userTokenCache,
    null);

var currentUser = (await app.GetAccountsAsync()).FirstOrDefault();
if (currentUser == null)
{
    throw new Exception("User was not stored properly");
}

// Acquire a token silently and force a new token to be created
var result = await app.AcquireTokenSilentAsync(
    _b2CScopes,
    currentUser,
    "",
    true
);
```

**Expected behavior**
The userTokenCache should be specific for each user and should not store other users tokens. It should also properly persist a users tokens attached to their session. The ConfidentialClientApplication should not have access to other accounts other than the logged in users.

**Actual behavior**
What happens is that if I log in first as one user and then again as a second user, the MsalSessionCache does not appear to properly retrieve information from the session on load.
```csharp
private void Load()
{
    SessionLock.EnterReadLock();
    _cache.Deserialize(_httpContext.Session.Get(_cacheId));
    SessionLock.ExitReadLock();
}
```
The session get with the unique to the user cache does not return anything. The TokenCache then gets filled out with the token from the first user and the second user.
When the code gets the first user (from the examples) it get the token based on the first logged in user, not the second user who is in the current context.

```csharp
// returns the first user, not the second
var currentUser = (await app.GetAccountsAsync()).FirstOrDefault();
```

I cannot tell if the issue has to do with how the cache is persisted, a problem with the token cache or a problem with how the ConfidentialClientApplication keeps track of accounts.

**Possible Solution**
I was able to move past this by searching through the applications account and matching based on the logged in users identifier. This fix does not work with the 2.6.0-preview as the list of accounts is zero (probably because the MsalSessionCache does not appear to properly persist).

```csharp
var signedInUserId = HttpContext.User.FindFirst(AzureAdB2COptions.ObjectIdentifierClaim)?.Value;
var idAndPolicy = signedInUserId + "-" + _azureAdB2COptions.SignUpSignInPolicy;
var azureClientApplication = GetConfidentialClientApplication(signedInUserId);
// Find the right user in the token cache using the user id and policy
var currentUser = (await azureClientApplication.GetAccountsAsync()).FirstOrDefault(user =>
    Equals(user.HomeAccountId.ObjectId.ToLower(), idAndPolicy.ToLower()));
```

**Comment by Jmaharman** [comment]
I've found the same issue / confusion. Bearing in mind the TokenCache is a new instantiation each time, I wouldn't expect to have previous tokens in there. On top of that, the implementations that wrap it, such as MsalSessionCache, all make it look like you would only be serializing the current user's token, not all tokens in memory from the point that the server was started.

I've looked at this for hours wondering what am I missing, in what other way is this meant to be used so that it makes sense.

Can we get some clarification on why it has been created in this way, what issues is it solving, and how are we meant to create long living caches without storing lots of other user details in the one user's cache.

Thanks

**Comment by jmprieur** [comment]
There seem to be a problem with the B2C Web samples. I need to have a look at them and fix them.
Confidential client applications, should managed one user token cache per user, keyed by the user unique ID, as shown in this sample: 
https://github.com/Azure-Samples/active-directory-aspnetcore-webapp-openidconnect-v2/blob/27cb318134f88fe2004f3932f06995b2f0195881/Extensions/InMemoryTokenCacheHelper.cs#L59-L64

````CSharp
public TokenCache GetCache(HttpContext httpContext, ClaimsPrincipal claimsPrincipal, AuthenticationProperties authenticationProperties, string signInScheme)
{
 string userId = claimsPrincipal.GetMsalAccountId();
 helper = new InMemoryTokenCacheHelper(userId, httpContext, memoryCache);
 return helper.GetMsalCacheInstance();
}
````

Note that the key is the GetMsalAccountId(), which retrieves a unique id for the user, and which is used as a key to the token cache:

```CSharp
public static string GetMsalAccountId(this ClaimsPrincipal claimsPrincipal)
{
 string userObjectId = GetObjectId(claimsPrincipal);
 string tenantId = GetTenantId(claimsPrincipal);


 if (string.IsNullOrWhiteSpace(userObjectId)) 
   throw new ArgumentOutOfRangeException("Missing claim 'http://schemas.microsoft.com/identity/claims/objectidentifier' or 'oid' ");

if (string.IsNullOrWhiteSpace(tenantId))
 throw new ArgumentOutOfRangeException("Missing claim 'http://schemas.microsoft.com/identity/claims/tenantid' or 'tid' ");

            string accountId = userObjectId + "." + tenantId;
            return accountId;
}
```

MSAL.NET was designed to fulfill a number of [scenarios](https://aka.ms-msal-net-scenarios) (public clients: where there are several accounts in the cache, and confidential clients where the Web App / Web API **needs to ensure** that each user gets its own cache serialized separately

I'll update the conceptual doc as well.

**Comment by Jmaharman** [comment]
Thanks for the quick response @jmprieur, I'd be interested to see how you fix the problem.

Personally I think once you look into fixing it, you'll find the same issue myself and @jondmcelroy have come across.

I was finding that even a completely newly instantiated TokenCache, with no state deserialized, had the previous users accounts still in the cache. I also tried to reload the state with a null for deserialize still failed to reset those users already in the cache, but no joy.

**Comment by jmprieur** [comment]
@Jmaharman : are you telling me that using the approach I proposed (One token cache per account ID), you still see  several accounts in the token cache?
I'll look at that

cc: @henrik-me

**Comment by Jmaharman** [comment]
That is what I was suggesting, however in modifying the sample to prove my point I've not been able to re-produce.

I will come back to you when I've got a concrete example, or if I find where I was going wrong.

**Comment by jmprieur** [comment]
ok. thanks @Jmaharman. Looking forward to having repro steps.

**Comment by Jmaharman** [comment]
Sorry for the delay. In looking closer I noticed a colleague set the project up with a preview version of the Microsoft.Identity.Client nuget package. Getting the latest version resolved the issue I described above.

No more problems here.

**Comment by jmprieur** [comment]
Thanks for the update, @Jmaharman !
Closing the issue since all is good.

**Comment by jondmcelroy** [comment]
I just tested and it appears the token cache is now no longer persisting even towards new instances with the newest version.
Problem can be reproed on 2.3, but is fixed by 2.6 at least. Thanks.
